Basic Unit tests for the `mig/shared/auth.py` module.

jonasbardino opened the pull request:
Unit tests for the `mig/shared/auth.py` module.

jonasbardino committed:
Basic Unit tests for the `mig/shared/auth.py` module.

jonasbardino committed:
Refactor cookie init into shared helper and expand test coverage.

jonasbardino committed:
Extended coverage with multiple corner-cases. Some tests are disabled because
they either require additonal error handling in the auth module or fail when
run as privileged user in the default docker CI.

albu-diku commented on tests/test_mig_shared_auth.py:
```diff
@@ -0,0 +1,544 @@
+        self.assertEqual(session_data.get('user_addr', ''), 'UNKNOWN',
+                         'should handle malformed session files gracefully')
+
+    @ unittest.skipIf(os.getuid() == 0, "Permissions don't work for priv users")
```
I am a little confused by this so asking for clarification - what circumstances lead to the necessity of this check? i.e. when are we running with privilege enough this does not work?

jonasbardino replied:
`make test` currently runs our unit tests as **root** inside the docker container and that effectively makes all POSIX ACL checks moot. Only manual local python runs or `make unittest` as an ordinary user will actually result in such checks working.

jonasbardino replied:
I can see we run `make unittest` in the CI containers (https://github.com/ucphhpc/migrid-sync/blob/next/.github/workflows/ci.yml), so the test **does** actually run there. 
I guess I must have noticed the issue in the first place when running `make test` locally and ending up as root in the resulting container. I only (re)discovered `make unittest` later.

Do you want to look into switching to an unprivileged user in the containers running the unit tests from `make test` or who is "we" in your previous comment? :-)
Anyway, since it _only_ affects local test runs and not CI I'd say we can merge and postpone that local `make test` fix for follow-up.

albu-diku commented on tests/test_mig_shared_auth.py:
```diff
@@ -0,0 +1,544 @@
+            auth.get_twofactor_secrets(self.configuration, TEST_USER_DN)
+        self.assertIn('failed to reset 2FA key', str(cm.exception))
+
+    @ unittest.skip("TODO: implement locking and enable")
```
Something weird going on with decorators, the space after the @ seems to be in this file multiple times.

jonasbardino replied:
fixed

albu-diku reviewed (changes requested):
The tests work and seem to cover session handling. But, another pass is needed over test naming and to better target the tests.

On a slightly broader level, I previously raised a flag about using both test naming and docstrings. This came up in one of the comments to a specific test, but am noting it specifically at the top level because we need to agree a norm about this to avoid confusion.

jonasbardino commented on tests/test_mig_shared_auth.py:
```diff
@@ -0,0 +1,544 @@
+        self.assertTrue(result, 'valid token should be accepted')
+
+    def test_verify_twofactor_token_invalid(self):
+        """Test invalid token rejection"""
```
I don't think unit tests are really that different to mandate special conventions from the general code base standard of using docstrings for all modules, classes plus for any non-trivial functions/methods. 
For basic tests it is _sometimes_ possible to come up with short and telling names, but we very quickly run into cases where it can simply not be summarized what a test does and why it is even relevant to test in less than the max line length characters, and then a docstring is needed anyway.

I've been lenient about demanding unit test method docstrings in reviews, but if you want a clearer convention I'd say always add docstrings even if just boilerplate ones like we have elsewhere in unit tests.

Of course I agree that they need to not be contradicting the method name and that these may need polish.

Good point about avoiding the duplication in the test class grouping and individual methods. I guess it comes from refactoring and I'll take another look.

jonasbardino replied:
On a closer look the _twofactor_ duplication comes from including the name of the tested auth function in the test method names. E.g. `test_twofactor_available` runs a test of the `twofactor_available` function, so it's not so strange.

jonasbardino replied:
Yes, you can use method name in filters.

I think it would be really ugly and unwieldy to use `test_number` and don't really see why we should not be able to go for test methods with brief telling names, _without_ trying to include every bit of info about what it does and why in it the name itself. That's what we do for every other function/method.
If you can completely convey what a test does in a brief method name then it's okay to leave out the docstring, but otherwise I think we're better off combining as described above.

jonasbardino committed:
Address various PR review feedback:
 - elaborate on strict_address session handling test in docstring
 - remove space after `@` in decorators
 - elaborate on `test_custom_interval_fallback` in docstring and assert error
 - elaborate on `test_session_file_permissions` in docstring
 - rename `test_session_file_permissions` to `.._file_access_restriction`

Additionally adjustments:
 - switch two stray f-string constructs to use our usual percent-string build
 - fix the `skip` decorator on `test_session_file_permissions` to always skip
   until the tested function properly handles permissions as it turned out that
   bit was really dependent on whatever shell umask was set.
 - fix `test_session_file_permissions` to actually work if enabled and tested
   function were to be extended to handles permissions as suggested.

jonasbardino commented on tests/test_mig_shared_auth.py:
```diff
@@ -0,0 +1,544 @@
+            self.assertEqual(session_data, {},
+                             'should return empty data for locked files')
+
+    def test_custom_interval_fallback(self):
```
This actually illustrates why I don't think it makes much sense to try to cram the test description into the method name. We'll quickly end up with extremely long and clunky names easily exceeding our default maximum line length even without indent and args. A short meaningful name and a doc string to explain details is much more readable, IMO.

jonasbardino replied:
Good, I adjusted the doc string a bit to clarify that is what it should do on invalid interval.

jonasbardino commented on tests/test_mig_shared_auth.py:
```diff
@@ -0,0 +1,544 @@
+        self.assertTrue(result, 'should accept token with custom interval')
+
+    def test_strict_address_session_handling(self):
+        """Test session handling with strict address enforcement"""
```
Further explained in docstring now. E.g. on GDP-mode sites like our SIF all SFTP/WebDAVS access is **only** accepted from the same source address as the client used to initiate the 2FA session. As a way to require 2FA for those services. This test verifies the usual 2FA session flow with just that added strict_address requirement.

jonasbardino committed:
Spin off individual 'atomic' partial tests to complement the full lifecycle
(integration) test as suggested in PR review.

jonasbardino commented on tests/test_mig_shared_auth.py:
```diff
@@ -0,0 +1,544 @@
+        environ = self._mimic_cookie_init(session_key)
+
+        # Expire session
+        expire_result = auth.expire_twofactor_session(self.configuration,
```
Good point :+1:  
I've added individual smaller `save`, `list` and `load` tests to help assure that each component works before continuing to this full lifecycle test, which is more of an integration test really.

NB: that link only works for you but I can reverse-engineer where to look on the internal wiki :-)

Sure, the possibilities for extending corner case tests are nearly endless.

albu-diku replied:
Yeah - one can think of it as needing to establish the the individual pieces of work before them all working in concert. From the viewpoint of fixing something, the issue with what you're calling a lifecycle test (nomenclature that I quite like, certainly as compared to integration which has a more specific use) would be the "what exactly broke" question - the smaller tests would help there.

Anyway, this looks good now though I'd have liked to see an isolated test for expiring a session (i.e. not relying on the library under test). Though I won't insist on it.

jonasbardino replied:
Yes, that makes sense. I'm all for further increasing test coverage, but slightly lean towards getting the basic tests in first.

albu-diku reviewed (approved):
Couple of unresolved conversation left purely for visibility - generally looks good now.

jonasbardino commented:
Just for the record I'll merge to get these unit tests in, but mark with the follow-up tag. There is a remaining general inconsistency to address with one or more unit tests getting skipped if run with `make test`, because it runs as root in the resulting container without ACL restrictions as described in comments above.
I suppose it requires the addition of an unprivileged user in the container launched from `envhelp/python3` and making sure the `envhelp/output/` etc from `make testconfig` is accessible to that same user. Perhaps that unprivileged user may even need to have the same UID as the active local user calling `make test`. It would not surprise me if that is not at all trivial ...
